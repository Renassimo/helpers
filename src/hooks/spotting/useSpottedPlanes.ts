import { useContext } from 'react';

import SpottingContext from '@/contexts/spotting';

const useSpottedPlanes = () => useContext(SpottingContext);

export default useSpottedPlanes;
