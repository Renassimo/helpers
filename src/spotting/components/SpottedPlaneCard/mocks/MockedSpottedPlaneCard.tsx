const MockedSpottedPlaneCard = jest.fn(
  ({ data, selectable }: { data: any; selectable?: any }) => {
    const { id, name, photoUrl, planespottersUrl, description, hashtags } =
      data;
    return (
      <div>
        Mocked Spotted Plane Card:{' '}
        {selectable ? 'selectable' : 'not selectable'}, id: {id}, name:
        {name}, photoUrl; {photoUrl}, planespottersUrl: {planespottersUrl},
        description: {description}, hashtags: {hashtags}
      </div>
    );
  }
);

export default MockedSpottedPlaneCard;
