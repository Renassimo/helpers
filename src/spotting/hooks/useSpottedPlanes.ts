import { useContext } from 'react';

import SpottingContext from '@/spotting/contexts/SpottingContext';

const useSpottedPlanes = () => useContext(SpottingContext);

export default useSpottedPlanes;
