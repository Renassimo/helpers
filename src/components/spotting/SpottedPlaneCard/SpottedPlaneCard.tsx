import React from 'react';

import { SpottedPlaneProviderData } from '@/types/spotting';

import CardContent from '@mui/material/CardContent';
import CardMedia from '@mui/material/CardMedia';
import CardActions from '@mui/material/CardActions';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import Button from '@mui/material/Button';
import LoadingButton from '@mui/lab/LoadingButton';

import OpenInNewIcon from '@mui/icons-material/OpenInNew';
import CollectionsIcon from '@mui/icons-material/Collections';
import ContentPasteIcon from '@mui/icons-material/ContentPaste';

import useSpottedPlanes from '@/hooks/spotting/useSpottedPlanes';
import useApplySpottedPlanes from '@/hooks/spotting/useApplySpottedPlanes';

import SelectableCard from '@/components/spotting/SelectableCard';
import NewFirstFlightForm from '@/components/spotting/NewFirstFlightForm';
import SpottedPlaneForm from '@/components/spotting/SpottedPlaneForm';

const SpottedPlaneCard = ({
  data,
  selectable = false,
}: {
  data: SpottedPlaneProviderData;
  selectable?: boolean;
}) => {
  const {
    id,
    name,
    photoUrl,
    planespottersUrl,
    description,
    hashtags,
    firstFlight,
    photosUrl,
    url,
  } = data;

  const { update, loading } = useApplySpottedPlanes();

  const {
    selectedIds,
    addSelectedId,
    removeSelectedIds,
    generateDescription,
    generateHashtags,
    clearHashtags,
    clearDescription,
  } = useSpottedPlanes();
  const isAnySelected = selectedIds.length > 0;
  const isSelected = selectedIds.includes(id);

  const toggleSelect = () => {
    if (isSelected) {
      removeSelectedIds([id]);
    } else {
      addSelectedId(id);
    }
  };

  const isPhotoSelectable = selectable && isAnySelected;

  const onPhotoClick = () => {
    if (isPhotoSelectable) toggleSelect();
  };

  const createDescription = () => {
    generateDescription(id);
    generateHashtags(id);
  };

  const apply = async () => {
    await update([data]);
  };

  const discard = () => {
    clearDescription(id);
    clearHashtags(id);
  };

  const showDescription = !!(description || hashtags);

  return (
    <SelectableCard
      isAnySelected={isAnySelected}
      selectable={selectable}
      selected={isSelected}
      toggleSelect={toggleSelect}
    >
      {
        <CardMedia
          image={photoUrl ?? '/images/plane.png'}
          title={name ?? id}
          sx={{ height: 140, cursor: isPhotoSelectable ? 'pointer' : '' }}
          onClick={onPhotoClick}
        />
      }
      <CardContent>
        <Typography gutterBottom variant="h6" component="h2">
          {name}
          <IconButton size="small" href={url} target="_blank" rel="noreferrer">
            <ContentPasteIcon fontSize="small" />
          </IconButton>
          {photosUrl && (
            <IconButton
              size="small"
              href={photosUrl}
              target="_blank"
              rel="noreferrer"
            >
              <CollectionsIcon fontSize="small" />
            </IconButton>
          )}
          {planespottersUrl && (
            <IconButton
              size="small"
              href={planespottersUrl}
              target="_blank"
              rel="noreferrer"
            >
              <OpenInNewIcon fontSize="small" />
            </IconButton>
          )}
        </Typography>
        {!firstFlight && (
          <NewFirstFlightForm id={data.id} disabled={showDescription} />
        )}
        {showDescription && <SpottedPlaneForm data={data} />}
      </CardContent>
      <CardActions sx={{ justifyContent: 'end' }}>
        {showDescription ? (
          <>
            <Button size="small" onClick={discard} disabled={loading}>
              Discard
            </Button>
            {selectable && (
              <LoadingButton size="small" onClick={apply} loading={loading}>
                Apply
              </LoadingButton>
            )}
          </>
        ) : (
          <Button size="small" onClick={createDescription}>
            Create description
          </Button>
        )}
      </CardActions>
    </SelectableCard>
  );
};

export default SpottedPlaneCard;
