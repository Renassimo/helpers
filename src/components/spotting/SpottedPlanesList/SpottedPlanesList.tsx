import React from 'react';

import useSpottedPlanes from '@/hooks/spotting/useSpottedPlanes';

import PlanesContainer from '@/components/spotting/PlanesContainer';
import SpottedPlaneCard from '@/components/spotting/SpottedPlaneCard';
import GroupPlanesModal from '@/components/spotting/GroupPlanesModal';

const SpottedPlanesList = () => {
  const { spottedPlanes } = useSpottedPlanes();

  return (
    <PlanesContainer>
      {spottedPlanes.map((item) => (
        <SpottedPlaneCard data={item} key={item.id} selectable />
      ))}
      <GroupPlanesModal />
    </PlanesContainer>
  );
};

export default SpottedPlanesList;
