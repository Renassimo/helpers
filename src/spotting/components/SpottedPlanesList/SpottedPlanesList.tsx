import React, { useState } from 'react';

import useSpottedPlanes from '@/spotting/hooks/useSpottedPlanes';

import PlanesContainer from '@/spotting/components/PlanesContainer';
import SpottedPlaneCard from '@/spotting/components/SpottedPlaneCard';
import GroupPlanesModal from '@/spotting/components/GroupPlanesModal';

const SpottedPlanesList = () => {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const { spottedPlanes } = useSpottedPlanes();

  return (
    <PlanesContainer>
      {spottedPlanes.map((item) => (
        <SpottedPlaneCard data={item} key={item.id} selectable />
      ))}
      <GroupPlanesModal
        isModalOpen={isModalOpen}
        setIsModalOpen={setIsModalOpen}
      />
    </PlanesContainer>
  );
};

export default SpottedPlanesList;
