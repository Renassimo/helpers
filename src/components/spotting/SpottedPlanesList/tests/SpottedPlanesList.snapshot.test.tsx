import renderWithTheme from '@/tests/helpers';

import useSpottedPlanes from '@/hooks/spotting/useSpottedPlanes';

import SpottedPlanesList from '@/components/spotting/SpottedPlanesList';
import SpottedPlaneCard from '@/components/spotting/SpottedPlaneCard';

import MockedSpottedPlaneCard from '@/components/spotting/SpottedPlaneCard/mocks';

jest.mock('@/hooks/spotting/useSpottedPlanes');
jest.mock('@/components/spotting/SpottedPlaneCard');

describe('SpottedPlanesList', () => {
  const mockedSpottedPlanes = [
    {
      id: 'plane-id',
      name: 'name',
      photoUrl: 'photoUrl',
      planespottersUrl: 'planespottersUrl',
      description: 'description',
      hashtags: '#hashtags',
    },
  ];
  const mockedUseSpottedPlanes = jest.fn(() => ({
    spottedPlanes: mockedSpottedPlanes,
  }));

  beforeEach(() => {
    (useSpottedPlanes as unknown as jest.Mock).mockImplementation(
      mockedUseSpottedPlanes
    );
    (SpottedPlaneCard as unknown as jest.Mock).mockImplementation(
      MockedSpottedPlaneCard
    );
  });

  test('renders successfully', () => {
    // Arrange
    // Act
    const { container } = renderWithTheme(<SpottedPlanesList />);
    // Assert
    expect(container).toMatchSnapshot();
    expect(MockedSpottedPlaneCard).toHaveBeenCalledWith(
      {
        data: mockedSpottedPlanes[0],
        selectable: true,
      },
      {}
    );
  });
});
