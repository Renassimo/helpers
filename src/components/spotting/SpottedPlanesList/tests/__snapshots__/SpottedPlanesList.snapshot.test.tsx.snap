// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpottedPlanesList renders successfully 1`] = `
<div>
  <div
    class="css-zsqo3r"
  >
    <div>
      Mocked Spotted Plane Card:
       
      selectable
      , id: 
      plane-id
      , name:
      name
      , photoUrl; 
      photoUrl
      , planespottersUrl: 
      planespottersUrl
      , description: 
      description
      , hashtags: 
      #hashtags
    </div>
  </div>
</div>
`;
